Introduce new query language (Approach 2) (#260)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

This specification defines the following new request parameters:

`presentation_definition`:
: A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorization Request.

`presentation_definition_uri`:

> : A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. See (#request_presentation_definition_uri) for more details. Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorizat
> ion Request.
> 

Reviewer: ```suggestion
: A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. See (#request_presentation_definition_uri) for more details. Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorization Request.

```

If the Verifier responds with any HTTP error response, the Wallet MUST terminate the process.

# Verifiable Presentation Query Language {#vp_query}

The Verifiable Presentation Query Language (VP Query) is a JSON-encoded query
language that allows the Verifier to request the Wallet to present Verifiable
Credentials that match the query. The query consists of a set of constraints
that the requested Verifiable Credentials must satisfy. The Wallet evaluates the
query against the Verifiable Credentials it holds and returns the Verifiable
Presentations that match the query.

A valid VP Query is defined as a JSON-encoded object with the following
top-level property:

> `expect_credentials`:

Reviewer 2: @martijnharing You can have either DNF or CNF and you need both to cater for all possible combinations of credential request.  "give me A or (B and C)"  can be done, if you have DNF.

> `expect_credentials`:

Author: The syntax proposed here is sufficient to express what you want. But I also tried to come up with an option that combines some of the elements of this proposal and a CNF in a new PR: #266 - please review!

in a group listed in `alternative_credentials` MUST NOT itself contain an
`alternative_credentials` property.

Each entry MUST be an object with the following properties:

`group`:
: OPTIONAL. A string identifying a group of claims for expressing alternatives
between claims. Multiple claims MAY have the same group identifier. The group
identifier for claims is not reflected in the response.

`path`:
: REQUIRED if the credential format uses a JSON-based claims structure; MUST NOT
be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
within the Verifiable Credential, as defined in (#claims_path_pointer).

`namespace`:
: REQUIRED if the credential format is based on ISO 18013-5; MUST NOT be present otherwise. 
The value MUST be a string that specifies the namespace of the claim
within the Verifiable Credential, e.g., `org.iso.18013.5.1`.

`claim_name`:
: REQUIRED if the credential format is based on ISO 18013-5; MUST NOT be present otherwise. 
The value MUST be a string that specifies the name of the claim within the provided namespace
in the Verifiable Credential, e.g., `first_name`.

> `value`:

Reviewer 3: Matching of values is a tricky subject. Some of the issues:

- Matching on values can be complicated for wallet / credential selection. Because you may not have access to the value of the element at the selection time.
- Even after wallet selection, having access to the values of the data elements may still only be available after user consent because of security reasons (decryption etc). Meaning that the presentation may fail only after user consent.
- Values of data elements are not just strings, integers or booleans. Depending on the credential format they can be maps, arrays, byte strings  or other items.
 
So besides requiring a stricter definition, mostly it's a mechanism that may just not be available because of security or ux reasons, so I would not be in favor of having it as an option since it's not a mechanism RPs can rely on.

Reviewer 4: It is a tricky subject, but the working group did spent a significant amount of time analysing what is needed from a query language for various use cases, and the conclusion was that matching of values is something that must be supported by a query language: https://github.com/openid/OpenID4VP/issues/144

It's also something supported by OID4VP today in presentation exchange, so something that is likely to be required if we want to move back to having only one query language in the future.

I acknowledge the difficulties, but I think we need to focus on how it can be done (and not focus on whether it should be done or not). Whether it's something that's mandatory to implement or not is I think something that's more likely to be dictated by the ecosystems. I understand some of the mandated EU use cases will require matching.

Author: We have syntax for other types of matching in the [Advanced Syntax for Claims specification](https://openid.bitbucket.io/ekyc/openid-connect-advanced-syntax-for-claims.html#name-defining-transformed-claims) which we could re-use here. There is at least one implementation of that spec, so we're not starting from scratch.

Author: Please see a proposal here: https://github.com/openid/OpenID4VP/pull/266#issuecomment-2362093202

: OPTIONAL. A string, integer or boolean value that specifies the expected value of the claim. If the
`value` property is present, the Wallet MUST return the claim only if the type and value
of the claim match the type and value specified in the query.

`values`:
: OPTIONAL. An array of strings, integers or boolean values that specifies the expected values of the claim.
If the `values` property is present, the Wallet MUST return the claim only if the
type and value of the claim both match for at least one of the elements in the array.

`alternative_claims`:
: OPTIONAL. An array of claim group identifiers that specify alternative groups of claims
that MAY be returned instead of the claim specified by the `path` or `claim_name` property if
that claim is not present or does not match the restriction expressed by `value` or `values`. Any
claim in a group listed in `alternatives` MUST NOT itself contain an `alternatives` property.

### Semantics of `alternative_claims` and `alternative_credentials`

By default, each element within `expect_credentials` or `expect_claims` MUST be returned, unless:

> - The element contains an `alternative_claims` or `alternative_credentials`

Reviewer 5: To me it feels a bit weird that alternative_claims is added to an expected claim. As this is a two way relation (you could also add it to the other claim). 

I think separating the requirements from the credentials is one of the things that really make sense to me in PEX, as it gives a nice structure:
- define which credentials can be presented.
- define which combination of credentials are allowed.

I feel this syntax intertwines it more and loses readability over it, while also making it quite hard to spot invalid configurations (as a claim/cred referenced as alternative can't reference alternatives itself). Would be interesting to see the complexity of implementing such an approach (hard to wrap my head around at the moment)

`kbjwt_signing_alg_values`:
: OPTIONAL. An array of strings that specifies the
allowed algorithms for signing the Key-Binding JWT (KB-JWT) (i.e., the part signed
by the holder).

`sd_alg_values`:
: OPTIONAL. An array of strings that specifies the allowed hash algorithms for
digests in the SD-JWT and KB-JWT.

### Format `mso_mdoc` {#format_mso_mdoc}

TBD

### Format `jwt_vp*`

TBD

## Claims Path Pointer {#claims_path_pointer}

A claims path pointer is a pointer into the JSON structure of the Verifiable
Credential, identifying one or more claims. A claims path pointer MUST be a
non-empty array of strings and non-negative integers. A `*` string value indicates
that all elements of the currently selected array(s) are to be selected; any other
string indicates that the respective key is to be selected; and a non-negative
integer indicates that the respective index in an array is to be selected. The path
is formed as follows:

  - Start with an empty array.
  - To address a particular claim within an object, append the key (claim name)
    to the array.
  - To address an element within an array, append the index to the array (as a
    non-negative, 0-based integer).

>   - To address all elements within an array, append the string `*` to the array.

Reviewer 5: With PEX it's possible to query for a specific item in an array based on value. Would that be possible?

I.e. i need to prove my german nationality, but my other nationalities are not important to disclose. This is something SD-JWT disclosures can support, but may be complex to convey in a vp query

Author: Please see a proposal here: https://github.com/openid/OpenID4VP/pull/266#issuecomment-2362093202

If the Verifier responds with any HTTP error response, the Wallet MUST terminate the process.

# Verifiable Presentation Query Language {#vp_query}

The Verifiable Presentation Query Language (VP Query) is a JSON-encoded query
language that allows the Verifier to request the Wallet to present Verifiable
Credentials that match the query. The query consists of a set of constraints

> that the requested Verifiable Credentials must satisfy. The Wallet evaluates the

Reviewer: ```suggestion
that the requested Verifiable Credentials MUST satisfy. The Wallet evaluates the
```

query against the Verifiable Credentials it holds and returns the Verifiable
Presentations that match the query.

A valid VP Query is defined as a JSON-encoded object with the following
top-level property:

`expect_credentials`:
: REQUIRED. An array of Credential Queries that specify the requested
Verifiable Credentials, as defined in (#credential_query).

Note: While this specification does not define additional top-level properties,
future extensions MAY define additional properties that modify the behavior of
the Wallet when processing the query.

## Credential Query {#credential_query}

A Credential Query is an object representing a request for one specific

> credential. It contains the following properties:

Reviewer: ```suggestion
Credential. It contains the following properties:
```

`group`:

> : REQUIRED. A string identifying the credential in the response, or, if multiple

Reviewer: ```suggestion
: REQUIRED. A string identifying the Credential in the response, or, if multiple
```

> credentials carry the same group identifier, the group of credentials that the

Reviewer: ```suggestion
Credentials carry the same group identifier, the group of Credentials that the
```

> requested credential belongs to.

Reviewer: ```suggestion
requested Credential belongs to.
```

`format`:
: REQUIRED. A string that specifies the format of the requested
Verifiable Credential. Valid Credential format identifier values are defined in
Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
supported Credential formats as defined in the Wallet's metadata.

> `expect_meta`: 
> : OPTIONAL. An object defining additional properties requested by the Verifier that
> apply to the metadata and validity data of the credential. The properties of
> this object are defined per format in (#format_specific_properties). If omitted,
> no specific restrictions are placed on the metadata or validity of the requested
> Credential.

Reviewer: ```suggestion
`expect_meta`: 
: OPTIONAL. An object defining additional properties requested by the Verifier that
apply to the metadata and validity data of the Credential. The properties of
this object are defined per format in (#format_specific_properties). If omitted,
no specific restrictions are placed on the metadata or validity of the requested
Credential.
```

`expect_claims`:
: OPTIONAL. An array that specifies the claims that the requested

> Verifiable Credential must contain. Each entry in the array corresponds to exactly

Reviewer: ```suggestion
Verifiable Credential MUST contain. Each entry in the array corresponds to exactly
```

I hope that `requested_claims` sounds better to you too and that we might evaluate this change as well.

> one claim that is to be returned by the Wallet to fulfill the credential

Reviewer: ```suggestion
one claim that is to be returned by the Wallet to fulfill the Credential
```

request. 

`alternative_credentials`:

> : OPTIONAL. An array of credential group identifiers that specify alternative

Reviewer: ```suggestion
: OPTIONAL. An array of Credential group identifiers that specify alternative
```

> groups of credentials that MAY be returned instead of the credential. Any credential

Reviewer: ```suggestion
groups of Credentials that MAY be returned instead of the Credential. Any Credential
```

in a group listed in `alternative_credentials` MUST NOT itself contain an
`alternative_credentials` property.

Each entry MUST be an object with the following properties:

`group`:
: OPTIONAL. A string identifying a group of claims for expressing alternatives
between claims. Multiple claims MAY have the same group identifier. The group
identifier for claims is not reflected in the response.

`path`:

> : REQUIRED if the credential format uses a JSON-based claims structure; MUST NOT

Reviewer: ```suggestion
: REQUIRED if the Credential format uses a JSON-based claims structure; MUST NOT
```

be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
within the Verifiable Credential, as defined in (#claims_path_pointer).

`namespace`:

> : REQUIRED if the credential format is based on ISO 18013-5; MUST NOT be present otherwise. 

Reviewer: ```suggestion
: REQUIRED if the Credential format is based on ISO 18013-5; MUST NOT be present otherwise. 
```

The value MUST be a string that specifies the namespace of the claim
within the Verifiable Credential, e.g., `org.iso.18013.5.1`.

`claim_name`:

> : REQUIRED if the credential format is based on ISO 18013-5; MUST NOT be present otherwise. 

Reviewer: ```suggestion
: REQUIRED if the Credential format is based on ISO 18013-5; MUST NOT be present otherwise. 
```

### Semantics of `alternative_claims` and `alternative_credentials`

By default, each element within `expect_credentials` or `expect_claims` MUST be returned, unless:

- The element contains an `alternative_claims` or `alternative_credentials`
  property and for at least one of the groups listed as an alternative, all
  elements in the group can be returned.
- The element is part of a group that is itself listed as an alternative in
  another element and returning the element is not needed to fulfill the
  previous rule.

Groups of claims are scoped to the specific credential; i.e., groups with the
same identifier that appear in separate Credential Query in `expect_credentials`
have no relation to each other.

## Format-specific Properties {#format_specific_properties}

The following format-specific properties are defined:

### Format `vc+sd-jwt` {#format_vc_sd_jwt}

`vct_values`:
: OPTIONAL. An array of strings that specifies allowed values for
the type of the requested Verifiable Credential. All elements in the array MUST
be valid type identifiers as defined in [@!I-D.ietf-oauth-sd-jwt-vc]. The Wallet

> MAY return credentials that inherit from any of the specified types, following

Reviewer: ```suggestion
MAY return Credentials that inherit from any of the specified types, following
```

the inheritance logic defined in [@!I-D.ietf-oauth-sd-jwt-vc].

`credential_signing_alg_values`:
: OPTIONAL. An array of strings that specifies
the allowed algorithms for signing the SD-JWT (i.e., only the issuer-signed part).

`kbjwt_signing_alg_values`:
: OPTIONAL. An array of strings that specifies the
allowed algorithms for signing the Key-Binding JWT (KB-JWT) (i.e., the part signed

> by the holder).

Reviewer: ```suggestion
by the Holder).
```

`sd_alg_values`:
: OPTIONAL. An array of strings that specifies the allowed hash algorithms for
digests in the SD-JWT and KB-JWT.

### Format `mso_mdoc` {#format_mso_mdoc}

> TBD

Reviewer: I got a very bad reaction from @bc-pi against the use of TBD in the specs, even if only draft. This is educating me.

Note: The `nonce` and `aud` are set to the `nonce` of the request and the Client Identifier of the Verifier, respectively, in the same way as for the Verifier, Verifiable Presentations to prevent replay.

# Examples for VP Queries {#vp_query_examples}

The following is a non-normative example of a VP Query that requests a Verifiable
Credential in the format `mso_mdoc` with the claims `vehicle_holder` and
`first_name`:

<{{examples/query_lang/simple_mdoc.json}}

The following is a non-normative example of a VP Query that requests the claim `last_name`, `date_of_birth`, and `email` and

- either the claim `postal_code`, or
- the claims `locality` and `region`:

<{{examples/query_lang/claims_alternatives.json}}

The following is a non-normative example of a VP Query that requests multiple

> Verifiable Credentials; all of them must be returned:

Reviewer: ```suggestion
Verifiable Credentials; all of them MUST be returned:
```

<{{examples/query_lang/multi_credentials.json}}

The following shows a complex query where the Wallet can either deliver

> the first credential (here called `pid`, containing all data the Verifier

Reviewer: ```suggestion
the first Credential (here called `pid`, containing all data the Verifier
```

> needs), or a second credential (here called `other_pid`), or present two

Reviewer: ```suggestion
needs), or a second Credential (here called `other_pid`), or present two
```

> credentials that together contain all the data the Verifier needs:

Reviewer: ```suggestion
Credentials that together contain all the data the Verifier needs:
```
